luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -12,484 +12,229 @@ function gadget:GetInfo()
 	end
 end
-TimedDamageDyingUnitsNames = nil
 
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+local function UpdateTimedAreas(frame)
+	local frameAreas = areas[frame % 30]
+	local frameRate = Spring.GetGameSpeed()
+	local frameTime = Spring.GetGameSeconds() - 0.5 / frameRate
+	-- local damageRate = 30 / frameRate -- todo
 
--- local aliveExplosions = {}
+	local explosion = {
+		weaponDef          = 0, -- todo: Great. Now I gotta make a weapon for all of these.
+		edgeEffectiveness  = 1,
+		explosionSpeed     = 10000,
+		damageGround       = false,
+	}
 
--- local function getRandomFreeExplosionID()
---     local attempts = 0
---     repeat
---         attempts = attempts + 1
---         local number = math.random(1,100000)
---         if not aliveExplosions[number] then
---             return number
---         end
---     until attempts >= 100000
---     return nil
--- end
+	-- Deal recurring damage by spawning explosions.
+	for index = #frameAreas, 1, -1 do
+		local timedArea = frameAreas[index]
+		if timedArea ~= empty then
+			if timedArea.endTime > frameTime then
+				-- todo: Will napalm/acid destroy shields now? Troubling thought.
+				explosion.weaponDef = timedArea.params.atd_wdef
+				Spring.SpawnExplosion(timedArea.x, timedArea.y, timedArea.z, 0, 0, 0, explosion)
+			else
+				StopTimedArea(frameAreas, index)
+			end
+		end
+	end
+end
 
-local aliveExplosions = {}
-local aliveExplosionsCounter = 1
+---------------------------------------------------------------------------------------------------------------
 
 function gadget:Initialize()
-    for id, a in pairs(TimedDamageWeapons) do
-        Script.SetWatchExplosion(id, true)
-    end
+	for wdid, _ in pairs(WeaponAreaDefs) do
+		Script.SetWatchExplosion(wdid, true)
+	end
 end
 
-function gadget:Explosion(weaponDefID, px, py, pz, AttackerID, ProjectileID)
-    if TimedDamageWeapons[weaponDefID] then
-        local currentTime = Spring.GetGameSeconds()
-        if py <= math.max(Spring.GetGroundHeight(px, pz), 0) + TimedDamageWeapons[weaponDefID].range*0.5 then
-            aliveExplosions[aliveExplosionsCounter] = {
-                x = px,
-                y = math.max(Spring.GetGroundHeight(px, pz), 0),
-                z = pz,
-                endTime = currentTime + TimedDamageWeapons[weaponDefID].time,
-                damage = TimedDamageWeapons[weaponDefID].damage,
-                range = TimedDamageWeapons[weaponDefID].range,
-                ceg = TimedDamageWeapons[weaponDefID].ceg,
-                cegSpawned = false,
-                damageCeg = TimedDamageWeapons[weaponDefID].damageCeg,
-                resistance = TimedDamageWeapons[weaponDefID].resistance,
-            }
-            aliveExplosionsCounter = aliveExplosionsCounter + 1
-        end
-    end
+function gadget:Explosion(weaponDefID, px, py, pz, attackerID, projectileID)
```

Comment:
Gating should be moved here and made explicit, rather than the line `if not params then return end` in `StartTimedArea`.

`if not WeaponAreaDefs[weaponDefID] then return end`